### PR TITLE
ROX-25503: Policy as code Custom Resource API

### PR DESCRIPTION
**Pull request by c-du**
### Description

change me!

### User-facing documentation

- [ ] CHANGELOG is updated **OR** update is not needed
- [ ] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

### Testing and quality

- [ ] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag
- [ ] CI results are inspected

#### Automated testing

- [ ] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [ ] modified existing tests

#### How I validated my change

change me!

### CODE REVIEW

File: proto/api/v1/policy_service.proto [hunk 1]
```diff
@@ -126,6 +126,14 @@ message GetPolicyMitreVectorsResponse {
   repeated storage.MitreAttackVector vectors = 2;
 }
 
+message ExportAsCustomResourcesRequest {
```
Comment (clickboo) [inline proto/api/v1/policy_service.proto:129]: I will avoid the word Export. Rcommend using "SaveAs"

SaveAsCustomResourceRequest (resource: singular)

File: proto/api/v1/policy_service.proto [hunk 2]
```diff
@@ -126,6 +126,14 @@ message GetPolicyMitreVectorsResponse {
...
+  repeated string policy_ids = 1;
+}
+
+message ExportAsCustomResourcesResponse {
```
Comment (clickboo) [inline proto/api/v1/policy_service.proto:133]: SaveAsCustomResourceResponse

File: proto/api/v1/policy_service.proto [hunk 3]
```diff
@@ -235,4 +243,9 @@ service PolicyService {
       body: "*"
     };
   }
+
+  // ExportAsCustomResources exports the policies, designed by the policy ids, to be saved as custome resources
+  rpc ExportAsCustomResources(ExportAsCustomResourcesRequest) returns (stream ExportAsCustomResourcesResponse) {
```
Comment (clickboo) [inline proto/api/v1/policy_service.proto:248]: Needs rename to SaveAs

**Comment by kylape** [comment]
Even with the rename to "SaveAs" this looks like the export endpoint.  Could we just add e.g. a `format` field to the `ExportPoliciesRequest`?  Or could we put this rpc under `/v1/policies/custom-resource` or `/v1/policies/{id}/custom-resource`?

**Comment by c-du** [comment]
> Even with the rename to "SaveAs" this looks like the export endpoint. Could we just add e.g. a `format` field to the `ExportPoliciesRequest`? Or could we put this rpc under `/v1/policies/custom-resource` or `/v1/policies/{id}/custom-resource`?

We had a discussion over this API endpoint in last meet. It is recommended to use the export policy path for consistency. I am open to discussions for this. The ExportPoliciesResponse returns storage.Policy and rely on the grpc-gateway to translate it to json. We do not have such functions for CR or Yaml and hence we need to have it separate.

File: central/policy/service/service_impl.go [hunk 4]
```diff
@@ -68,6 +71,7 @@ var (
 			"/v1.PolicyService/ExportPolicies",
 			"/v1.PolicyService/PolicyFromSearch",
 			"/v1.PolicyService/GetPolicyMitreVectors",
+			"/v1.PolicyService/SaveAsCustomResources",
```
Comment (clickboo) [inline central/policy/service/service_impl.go:74]: SaveAsCustomResource (singular)?

File: central/policy/service/service_impl.go [hunk 5]
```diff
@@ -717,10 +721,10 @@ func (s *serviceImpl) ExportPolicies(ctx context.Context, request *v1.ExportPoli
 	if err != nil {
 		return nil, err
 	}
-	errDetails := &v1.ExportPoliciesErrorList{}
+	errDetails := &v1.PolicyOperationErrorList{}
 	for _, missingIndex := range missingIndices {
 		policyID := request.PolicyIds[missingIndex]
-		errDetails.Errors = append(errDetails.Errors, &v1.ExportPolicyError{
+		errDetails.Errors = append(errDetails.Errors, &v1.PolicyOperationError{
```
Comment (clickboo) [inline central/policy/service/service_impl.go:727]: Not sure, but confirming this does not break UI? or Scripts? Backward compatible?

File: central/policy/service/service_impl.go [hunk 6]
```diff
@@ -1042,7 +1046,7 @@ func flattenPolicyGroupMap(policyGroupMap map[string][]*storage.PolicyGroup) []*
 
 		var policyValueLists []*storage.PolicyValue
 		for _, policyGroup := range singleGroupList {
-			// For now we don't care which search term a policy value came from because no two search terms can
+			// For now we don't care which search term a policy value came f/rom because no two search terms can
```
Comment (clickboo) [inline central/policy/service/service_impl.go:1049]: revert?

File: central/policy/service/service_impl.go [hunk 7]
```diff
@@ -1122,3 +1126,52 @@ func combineStrings(toCombine [][]string) []string {
 	}
 	return nil
 }
+
+func (s *serviceImpl) SaveAsCustomResources(ctx context.Context, request *v1.SaveAsCustomResourcesRequest) (*v1.SaveAsCustomResourcesResponse, error) {
```
Comment (clickboo) [inline central/policy/service/service_impl.go:1130]: Singular resource everywhere?

File: proto/api/v1/policy_service.proto [hunk 8]
```diff
@@ -235,4 +243,9 @@ service PolicyService {
       body: "*"
     };
   }
+
+  // SaveAsCustomResources saves the policies, designed by the policy ids, as custome resources
+  rpc SaveAsCustomResources(SaveAsCustomResourcesRequest) returns (SaveAsCustomResourcesResponse) {
+    option (google.api.http) = {get: "/v1/policies/save-as"};
```
Comment (kylape) [inline proto/api/v1/policy_service.proto:249]: Shouldn't this be `post` instead of `get`?

File: central/policy/service/service_impl.go (hunk 5 above)
Comment (vjwilson) [reply central/policy/service/service_impl.go:727]: I think the API gets the errors as one concatenated string, the same one repeated in both the `message` and `error` field of an error response.

In short, I don't think the UI's usage of the API can know or care what type of error it is. Consumers of the API through gRPC _may_ be affected by this, however.

File: central/policy/service/service_impl.go (hunk 4 above)
Comment (c-du) [reply central/policy/service/service_impl.go:74]: I am following the same naming as other entries in this service. We do support save-as multiple policies.

File: central/policy/service/service_impl.go (hunk 6 above)
Comment (c-du) [reply central/policy/service/service_impl.go:1049]: I do not even know how I made this change. Thank you.

File: central/policy/service/service_impl.go [hunk 9]
```diff
@@ -1122,3 +1125,52 @@ func combineStrings(toCombine [][]string) []string {
...
+		errDetails.Errors = append(errDetails.Errors, &v1.PolicyOperationError{
+			PolicyId: policyID,
+			Error: &v1.PolicyError{
+				Error: "not found",
+			},
+		})
+		log.Warnf("A policy error occurred for id %s: not found", policyID)
+	}
+	if len(missingIndices) > 0 {
+		statusMsg, err := status.New(codes.InvalidArgument, "Some policies could not be retrieved. Check the error details for a list of policies that could not be found").WithDetails(errDetails)
```
Comment (clickboo) [inline central/policy/service/service_impl.go:1146]: Failed to retrieve all policies. Check error details for a list of policies that could not be retrieved.

File: central/policy/service/service_impl.go [hunk 10]
```diff
@@ -1122,3 +1125,52 @@ func combineStrings(toCombine [][]string) []string {
...
+				PolicyId: policy.GetId(),
+				Error: &v1.PolicyError{
+					Error: errors.Wrap(err, "Failed to marshal policy to custom resource").Error(),
+				},
+			})
+		}
+		resp.CustomResources = append(resp.CustomResources, cr)
+	}
+	if len(errDetails.GetErrors()) > 0 {
+		statusMsg, err := status.New(codes.InvalidArgument, "Some policies could not be marshal to custom resources. Check the error details for the list of policies").WithDetails(errDetails)
```
Comment (clickboo) [inline central/policy/service/service_impl.go:1168]: Failed to save all policies as custom resources. Check error details for a list of policies.

**Review by clickboo — APPROVED** [review]

**Review by kylape — COMMENTED** [review]
I did a manual validation that saving a CR using the `save-as` endpoint works and can be applied successfully to k8s (minus the `policyName` which is not synced up quite yet).

**Review by kylape — APPROVED** [review]
Sorry, I meant to approve :)

**Pull request merged**